grep/grep.1
'\" t
.\" Sccsid @(#)grep.1	1.36 (gritter) 8/14/05
.\" Parts taken from grep(1), Unix 7th edition:
.\" Copyright(C) Caldera International Inc. 2001-2002. All rights reserved.
.\"
.\" SPDX-Licence-Identifier: Caldera
.\"
.TH GREP 1 "8/14/05" "Heirloom Toolchest" "User Commands"
.SH NAME
grep \- search a file for a pattern
.SH SYNOPSIS
.HP
.ad l
.nh
\fB/usr/5bin/grep\fR [\fB\-bcHhilnrRsvwz\fR]
\fIpattern\fR [\fIfile\fR\ ...]
.HP
.PD 0
.ad l
\fB/usr/5bin/posix/grep\fR [\fB\-E\fR|\fB\-F\fR]
\fB\-e\fI\ pattern_list\fR\ ...
[\fB\-f\fI\ pattern_file\fR] [\fB\-c\fR|\fB\-l\fR|\fB\-q\fR]
[\fB\-bhinrRsvwxz\fR] [\fIfile\fR\ ...]
.HP
.ad l
\fB/usr/5bin/posix/grep\fR [\fB\-E\fR|\fB\-F\fR]
\fB\-f\fI\ pattern_file\fR
[\fB\-e\fI\ pattern_list\fR\ ...] [\fB\-c\fR|\fB\-l\fR|\fB\-q\fR]
[\fB\-bhinrRsvwxz\fR] [\fIfile\fR\ ...]
.HP
.ad l
\fB/usr/5bin/posix/grep\fR [\fB\-E\fR|\fB\-F\fR]
[\fB\-c\fR|\fB\-l\fR|\fB\-q\fR] [\fB\-bhinrRsvwxz\fR]
\fIpattern_list\fR [\fIfile\fR\ ...]
.br
.PD
.ad b
.hy 1
.SH DESCRIPTION
The
.B grep
command searches the lines of the specified files
(or of standard input)
for occurrences of the regular expression
.I pattern.
The default behavior is to print each matching line to standard output.
.PP
The
.B /usr/5bin/grep
command accepts one pattern
that is treated as a simple regular expression;
it uses a compact nondeterministic algorithm.
.PP
The
.B /usr/5bin/posix/grep
command uses basic regular expressions by default
and accepts a newline-separated list of patterns
as described for the
.B \-e
option below.
It uses a deterministic algorithm with moderate space requirements
for most expressions;
backreferences, word delimiters, and multi-character collating elements
cause a nondeterministic algorithm to be used.
.PP
.B /usr/5bin/s42/grep
and
.B /usr/5bin/posix2001/grep
are identical to
.BR /usr/5bin/posix/grep .
.PP
See the description of
.IR ed (1)
for the specifications of simple and basic regular expressions.
.PP
Care should be taken when using the characters
$ * [ ^ | ? \' " ( ) and \e in the expression
as they are also meaningful to the Shell.
It is safest to enclose the entire expression
argument in single quotes \' \'.
.PP
Both
.B /usr/5bin/grep
and
.B /usr/5bin/posix/grep
accept the following options:
.TP
.B \-b
Each line is preceded by the block number on which it was found.
This is sometimes useful
in locating disk block numbers by context.
Block numbers start with 0.
.TP
.B \-c
Only a count of matching lines is printed.
.TP
.B \-H
Prints the name of a input file before a match even if it's
just one.
.br
Useful when finding for matches in a directory tree with
.B find(1)
.B -exec
option switch.
.TP
.B \-h
Normally, the name of each input file is printed before a match
if there is more that one input file.
When this option is present, no file names are printed.
.TP
.B \-i
Upper- and lowercase differences are ignored when searching matches.
.TP
.B \-l
The names of files with matching lines are listed
(once) separated by newlines.
.TP
.B \-n
Each line is preceded by its line number in the file.
Line numbers start with 1.
.TP
.B \-s
Error messages for nonexistent or unreadable files are suppressed.
.TP
.B \-v
All lines but those matching are printed.
.PP
The following options are supported by
.B /usr/5bin/posix/grep
only:
.TP
.BI \-e\  pattern_list
Specifies one or more patterns, separated by newline characters.
A line is selected if one or more of the specified patterns are found.
.TP
.B \-E
All patterns are interpreted as extended regular expressions
as described in
.IR egrep (1).
.TP
.BI \-f\  pattern_file
One or more patterns, separated by newline
characters, are read from
.I pattern_file.
.TP
.B \-F
All patterns are interpreted as fixed strings,
as with
.IR fgrep (1).
.TP
.B \-q
Do not write anything to standard output.
.TP
.B \-x
Consider only lines consisting of the pattern as a whole,
like a regular expression surrounded by
.I ^
and
.I $.
.PP
The following options are supported as extensions:
.TP
.B \-r
With this option given,
.I grep
does not directly search in each given file that is a directory,
but descends it recursively
and scans each regular file found below it.
Device files are ignored.
Symbolic links are followed.
.TP
.B \-R
Operates recursively as with the
.I \-r
option,
but does not follow symbolic links that point to directories
unless if they are explicitly specified as arguments.
.TP
.B \-w
Searches for the patterns treated as words,
as if they were surrounded by `\e<\ \e>'.
Only available if neither the
.I \-E
nor the
.I \-F
option are also supplied.
.TP
.B \-z
If an input file is found to be compressed with
.IR compress (1),
.IR gzip (1),
or
.IR bzip2 (1),
the appropriate compression program is started,
and
.I grep
searches for the pattern in its output.
.SH "ENVIRONMENT VARIABLES"
.TP
.BR LANG ", " LC_ALL
See
.IR locale (7).
.TP
.B LC_COLLATE
Affects the collation order for range expressions,
equivalence classes, and collation symbols
in basic regular expressions.
.TP
.B LC_CTYPE
Determines the mapping of bytes to characters
in both simple and basic regular expressions,
the availability and composition of character classes
in basic regular expressions,
and the case mapping for the
.B \-i
option.
.SH "SEE ALSO"
ed(1),
egrep(1),
fgrep(1),
sed(1),
locale(7)
.SH DIAGNOSTICS
Exit status is 0 if any matches are found,
1 if none, 2 for syntax errors or inaccessible files.
.SH NOTES
If a line contains a
.SM NUL
character,
only matches up to this character are found
(unless
.B /usr/5bin/posix/grep
is used with the
.I \-F
option).
The entire matching line will be printed.
.PP
The LC_COLLATE variable has currently no effect.
Ranges in bracket expressions are ordered
as byte values in single-byte locales
and as wide character values in multibyte locales;
equivalence classes match the given character only,
and multi-character collating elements are not available.
.PP
The options supported by
.B /usr/5bin/posix/grep
that are not accepted by
.B /usr/5bin/grep
can easily be replaced by portable constructs:
Use
.I egrep
instead of
.BR \-E ,
.I fgrep
instead of
.BR \-F .
Use
.I egrep
if you need the
.B \-e
or
.B \-f
option,
use only one of them and that only once;
if necessary, use text processing tools
to generate a single expression list before.
Redirect standard output to
.I /dev/null
for
.B \-q
(the possible speedup with
.I \-q
is never worth human time spent with porting scripts),
and use the
.RI ` ^ '
and
.RI ` $ '
meta-characters instead of
.BR \-x .


CHANGES
Release 250209
* cpio: Fix manual page formatting.
* apply: Fixed the command string magic character computation algorithm that
  got garbled after a refactor made in the name of performance.
* build/mk.config, chroot, nawk: Adding -ltinfo for static compilation with
  glibc and making changes to adhere to C standards (patch by Gabriel Souza).
* [NEW] wall, write: New implementation ported from UNIX 7th (a.k.a. 32V),
  bulletproofed against buffer overflows and the CVE-2024-28085, which is
  available as the 'ssafe()' function at libcommon.
* rm: Alias '-e' flag to more known '-v' for verbose.
* date, shl: Fix compilation on platforms where (struct utmpx).ut_tv is not
  of type 'timeval'.
* rev: UTF-8 support, borrowed from suckless sbase. In varietate concordia.
* [NEW] apply, pick: Complete implementation for both apply and pick, from
  Research UNIX 8th and 10th respectively; extended payloads for pick in
  comparison with the original implementation.
* seq: Support for counting from negative numbers to positive or making a
  negative sequence.
* rev: Full refactor (made by Samuel Brederodes).
* Fixes for building dynamically on Arch Linux and other GNU C
  Library-based Linux systems which ships exclusively dynamic versions
  of its libraries, also remove `readlink_ucb' object file at `readlink/'
  build directory when executing `make clean' (patch by Molly A. McCollum).
* Remove `register's from legacy code that was living unchanged since at
  least the 80's, since modern linters started complaining about it and
  it is not used as much on most common architectures such as x86
  (patch by Samuel Brederodes).
* watch: `use_default_colors()', do not force black and white.
* [NEW] rev: New implementation, not capable of UTF-8 yet.
* cpio, diff, tabs: Fix duplicated definitions (made/reported by crumbtoo,
  merged by me, Luiz Antônio Rangel). Thanks! --- or better: Merci!
* rm: Added real support for the POSIX 1003.1(2008) standard, implemented
  IBM AIX(R)-style verbosing and 4.4BSD's '-d' option for directories ---
  rmdir(1)-esque payload --- as extensions to the standard.
* seq: Made its payload similar to UNIX v8's implementation, but with some
  extensions such as a fast algorithm for -w or -s --- which was already
  present before the (almost) full rewrite.
* timeout: avoid buffer overrun, also did some changes at the source code,
  such as cleaning up comments and initializing 'commandv' and 'fst_commandv'.
* watch: Used esr's deheader for removing unnecessary headers, also use
  pfmt(stderr, MM_ERROR, ...)  instead of prerror().
* watch, timeout: Some small corrections at the manual page.
* cpio: Skip over zip extra headers with fewer than four bytes (the minimum
  length for them to contain any information) instead of reporting a bad
  header and terminating.
* pg: Handle cases where standard output is a terminal but not open for
  reading (bug reported by Keith Doyle).
* cpio, tar: If an archive contains a file that is a hardlink to itself on
  extraction, do nothing at all instead of discarding its data. This mainly
  happens if a file with more than one link is included under the same name
  in an archive twice. Since special situations on extraction such as bind
  mounts on Linux may lead to this scenario even in cases where both link
  names are different, this is now treated like a successful link() step
  (bugreport by Thomas Graf).
* Made the getopt() substitute work with the getopt$UNIX2003() function
  redirection mechanism on Mac OS X Leopard.
* mknod, pgrep, ps: Mac OS X Leopard compilation fixes.
* w, whodo: On a 32-bit system where /proc/kcore is larger than 2 GB,
  avoid a harmless error message (bug reported by Pierre Gentile).
* ptime: Works on recent Linux kernels again.
* col: The behavior of the -x flag has been inverted to actually match
  the documentation, i.e. to suppress conversion of spaces to tabs (patch
  by Markus Schnalke).
* cut: With "-s", print the last field of a line that was also the last
  field specified with "-f" (bugreport by Pierre Gentile).

Release 240220
* [NEW] seq: Stone-simple implementation; needs some improvement yet.
* [NEW] watch: Complete, simple and stable implementation.
* [NEW] timeout: Complete implementation with extended payloads.
* readlink: Different help messages for both UCB and default variants.

Release 230503
* chroot: "docs: Update date at chroot.8 manual page."
* chroot: "chore: Remove comment, will move to GitHub"
* chroot: "docs: Rewrote chroot(8) manual page"
* chroot: "chore: Add ``usage()`` function at chroot.c"
* tabs: "fix: [t,g]roff formatting at tabs/fspec.5."
* readlink: "docs: Created new manual page for the non-UCB variant. (#24)"
* Makefile.mk: "fix: Don't symbolic/hard link readlink from UCBDIR to DEFDIR anymore"
* readlink: "docs: Add comments clarifying the C preprocessor directives."
* readlink: "fix: Ignore ``-n`` option when operating multiple files."
* readlink: "feat: Create non-UCB/GNU-alike variant."
* readlink: "fix: Made ``name_size`` private again, so ``pfmt()`` won't go segmentation fault."
* readlink: "feat: First step, start modularizing the code."
* libcommon: "fix: getdirentries replacement for OBSD >5.4"
* fix-me: For new releases, finish porting for OpenBSD and implement watch(1) and timeout(1).

Release 230305
* getconf: "Now _NPROCESSORS_ONLN and _NPROCESSORS_CONF are implemented."
* (*): "Removed the remaining licence headers, now every licence is referenced by its SPDX identifier."

Release 230228
* [NEW] readlink: New implementation, extensively based on OpenBSD's.
* readlink: It's installed at /usr/ucb per default and symbolic/hard linked to /usr/5bin.

Release 230120
* Notes for the library directories if linking against the GNU C library.
* rm: Consistency within standards: the SUS program shouldn't have the POSIX 1003.1(2008) fix, just the main program.
* Made $(ROOT) not hardcoded anymore.

grep/grep.c
/*
 * grep - search a file for a pattern
 *
 * Gunnar Ritter, Freiburg i. Br., Germany, April 2001.
 */
/*
 * Copyright (c) 2003 Gunnar Ritter
 * Copyright (C) 2023 Luiz Antônio Rangel
 * SPDX-Licence-Identifier: Zlib
 */

/*	Sccsid @(#)grep.c	1.53 (gritter) 12/27/06>	*/

/*
 * Code common to all grep flavors.
 */

#include	<sys/types.h>
#include	<sys/stat.h>
#include	<sys/mman.h>
#include	<sys/wait.h>
#include	<fcntl.h>
#include	<unistd.h>
#include	<stdio.h>
#include	<string.h>
#include	<stdlib.h>
#include	<libgen.h>
#include	<locale.h>
#include	<limits.h>
#include	<ctype.h>
#include	<dirent.h>
#include	<errno.h>

#include	"grep.h"
#include	"alloc.h"

/*
 * Generic flags and the like.
 */
int		Eflag;			/* use EREs */
int		Fflag;			/* use fixed strings */
int		bflag;			/* print buffer count */
int		cflag;			/* print count only */
int		fflag;			/* had pattern file argument */
int		Hflag;			/* print filenames along with the match */
int		hflag;			/* do not print filenames */
int		iflag;			/* ignore case */
int		lflag;			/* print filenames only */
int		nflag;			/* print line numbers */
int		qflag;			/* no output at all */
int	(*rflag)(const char *, struct stat *);	/* operate recursively */
int		sflag;			/* avoid error messages */
int		vflag;			/* inverse selection */
int		wflag;			/* search for words */
int		xflag;			/* match entire line */
int		zflag;			/* decompress compressed files */
int		mb_cur_max;		/* avoid multiple calls to MB_CUR_MAX */
unsigned	status = 1;		/* exit status */
off_t		lmatch;			/* count of line matches */
off_t		lineno;			/* current line number */
char		*progname;		/* argv[0] to main() */
char		*filename;		/* name of current file */
char		*options;		/* for getopt() */
void		(*build)(void);		/* compile function */
int		(*match)(const char *, size_t); /* comparison function */
int		(*range)(struct iblok *, char *); /* grep range of lines */

/*
 * Regexp variables.
 */
struct expr	*e0;			/* start of expression list */
enum matchflags	matchflags;		/* matcher flags */

/*
 * To avoid link loops with -r.
 */
static struct	visit {
	ino_t	v_ino;
	dev_t	v_dev;
} *visited;
static int	vismax;			/* number of members in visited */

/*
 * Lower-case a character string.
 */
size_t
loconv(char *dst, char *src, size_t sz)
{
	char	*odst = dst;

	if (mbcode) {
		char	mb[MB_LEN_MAX];
		wchar_t wc;
		int len, i, nlen;

		while (sz > 0) {
			if ((*src & 0200) == 0) {
				*dst++ = tolower(*src);
				src++;
				sz--;
			} else if ((len = mbtowc(&wc, src, sz)) <= 0 ||
					len > sz) {
				*dst++ = *src++;
				sz--;
			} else {
				wc = towlower(wc);
				if (len >= mb_cur_max) {
					if ((nlen = wctomb(dst, wc)) <= len) {
						dst += nlen;
						src += len;
						sz -= len;
					} else {
						*dst++ = *src++;
						sz--;
					}
				} else {
					if ((nlen = wctomb(mb, wc)) <= len) {
						sz -= len;
						src += len;
						for (i = 0; i < nlen; i++)
							*dst++ = mb[i];
					} else {
						*dst++ = *src++;
						sz--;
					}
				}
			}
		}
	} else {
		while (sz--) {
			*dst++ = tolower(*src & 0377);
			src++;
		}
	}
	return dst - odst;
}

/*
 * Determine if pat ends with an unescaped dollar sign.
 */
static int
termdollar(const char *pat, long len)
{
	int	dollar = 1;

	if (len == 0 || pat[len - 1] != '$')
		return 0;
	pat += --len - 1;
	while (len-- && *pat-- == '\\')
		dollar = !dollar;
	return dollar;
}

/*
 * Surround the pattern with \< \>.
 */
void
wcomp(char **pat, long *len)
{
	char	*wp = smalloc(*len + 5);

	memcpy(&wp[2], *pat, *len);
	if ((*pat)[0] == '^')
		memcpy(wp, "^\\<", 3);
	else
		memcpy(wp, "\\<", 2);
	if (termdollar(*pat, *len))
		strcpy(&wp[*len-1+2], "\\>$");
	else
		strcpy(&wp[*len+2], "\\>");
	*len += 4;
	*pat = wp;
}

static struct iblok *
redirect(struct iblok *ip, const char *arg0, const char *arg1)
{
	struct iblok	*nip = NULL;
	int	pd[2];
	pid_t	pid;

	if (pipe(pd) < 0)
		return NULL;
	switch (pid = fork()) {
	case 0:
		if (lseek(ip->ib_fd, -(ip->ib_end - ip->ib_cur),
					SEEK_CUR) == (off_t)-1) {
			int	xpd[2];
			if (pipe(xpd) == 0 && fork() == 0) {
				ssize_t	rd, wo, wt;
				close(xpd[0]);
				for (;;) {
					rd = ip->ib_end - ip->ib_cur;
					wo = wt = 0;
					do {
						if ((wo = write(xpd[1],
								&ip->ib_cur[wt],
								rd - wt))
								<= 0) {
							if (errno == EINTR)
								continue;
							_exit(0);
						}
						wt += wo;
					} while (wt < rd);
					if (ib_read(ip) == EOF)
						break;
					ip->ib_cur--;
				}
				_exit(0);
			} else {
				close(xpd[1]);
				dup2(xpd[0], 0);
				close(xpd[0]);
			}
		} else {
			if (ip->ib_fd)
				dup2(ip->ib_fd, 0);
		}
		if (ip->ib_fd)
			ib_close(ip);
		else
			ib_free(ip);
		dup2(pd[1], 1);
		close(pd[0]);
		close(pd[1]);
		execlp(arg0, arg0, arg1, NULL);
		fprintf(stderr, "%s: could not exec %s\n", progname, arg0);
		_exit(0177);
		/*NOTREACHED*/
	case -1:
		fprintf(stderr, "%s: cannot fork()\n", progname);
		status = 2;
		return NULL;
	default:
		close(pd[1]);
		nip = ib_alloc(pd[0], 0);
		nip->ib_pid = pid;
		return nip;
	}
}

/*
 * Report a matching line.
 */
void
report(const char *line, size_t llen, off_t bcnt, int addnl)
{
	if (filename && (!hflag || Hflag))
		printf("%s:\040", filename);
#ifdef	LONGLONG
	if (bflag)
		printf("%llu:", (long long)bcnt);
	if (nflag)
		printf("%llu:", (long long)lineno);
#else	/* !LONGLONG */
	if (bflag)
		printf("%lu:", (long)bcnt);
	if (nflag)
		printf("%lu:", (long)lineno);
#endif	/* !LONGLONG */
	if (line && llen)
		fwrite(line, sizeof *line, llen, stdout);
	if (addnl)
		putchar('\n');
}

/*
 * Check line for match. If necessary, the line gets NUL-terminated (so
 * its address range must be writable then). When ignoring character case,
 * a lower-case-only copy of the line is made instead. If a match is found,
 * statistics are printed. Returns 1 if main loop shall terminate, 0 else.
 */
static int
matchline(char *line, size_t sz, int putnl, struct iblok *ip)
{
	size_t	csz = sz;
	int terminate = 0;
	char lbuf[512], *abuf = NULL, *cline = line;

	if (iflag && (matchflags & MF_LOCONV)) {
		if (sz >= sizeof lbuf - 1) {
			abuf = smalloc(sz + 1);
			cline = abuf;
		} else
			cline = lbuf;
		csz = loconv(cline, line, sz);
		cline[csz] = '\0';
	} else if (matchflags & MF_NULTERM)
		cline[sz] = '\0';
	lineno++;
	if (match(cline, csz) ^ vflag) {
		lmatch++;
		if (qflag == 0) {
			if (status == 1)
				status = 0;
			if (lflag) {
				puts(filename ? filename : stdinmsg);
			} else if (!cflag)
				report(line, sz, (ib_offs(ip)-1) / BSZ, putnl);
		} else
			exit(0);
		if (qflag || lflag)
			terminate = 1;
	}
	if (abuf)
		free(abuf);
	return terminate;
}

/*
 * Check all lines within ip->ib_cur and last which contains the last
 * newline. If the main loop shall terminate, 1 is returned.
 */
static int
gn_range(struct iblok *ip, char *last)
{
	char *nl;

	while ((nl = memchr(ip->ib_cur, '\n', last + 1 - ip->ib_cur)) != NULL) {
		if (matchline(ip->ib_cur, nl - ip->ib_cur,  1, ip))
			return 1;
		if (nl == last)
			return 0;
		ip->ib_cur = nl + 1;
	}
	return 0;
}

/*
 * Main grep routine. The line buffer herein is only used for overlaps
 * between file buffer fills.
 */
static struct iblok *
grep(struct iblok *ip)
{
	char *line = NULL;		/* line buffer */
	char *lastnl;		/* last newline in file buffer */
	size_t sz = 0;			/* length of line in line buffer */
	char *cp;
	int hadnl;			/* lastnl points to newline char */
	int	oom = 0;		/* got out of memory */

	lineno = lmatch = 0;
	if (ib_read(ip) == EOF)
		goto endgrep;
	ip->ib_cur--;
	if (zflag) {
		struct iblok	*np;
		for (;;) {
			sz = ip->ib_end - ip->ib_cur;
			if (sz > 3 && memcmp(ip->ib_cur, "BZh", 3) == 0)
				np = redirect(ip, "bzip2", "-cd");
			else if (sz > 2 &&
					memcmp(ip->ib_cur, "\37\235", 2) == 0)
				np = redirect(ip, "zcat", NULL);
			else if (sz > 2 &&
					memcmp(ip->ib_cur, "\37\213", 2) == 0)
				np = redirect(ip, "gzip", "-cd");
			else
				break;
			if (np == NULL)
				break;
			if (ip->ib_fd)
				ib_close(ip);
			else
				ib_free(ip);
			ip = np;
			if (ib_read(ip) == EOF)
				goto endgrep;
			ip->ib_cur--;
		}
	}
	for (;;) {
		for (lastnl = ip->ib_end - 1;
				*lastnl != '\n' && lastnl > ip->ib_cur;
				lastnl--);
		if (hadnl = (ip->ib_cur < ip->ib_end && *lastnl == '\n'))
			if (range(ip, lastnl))
				break;
		if (lastnl < ip->ib_end - hadnl) {
			/*
			 * Copy the partial line from file buffer to line
			 * buffer. Allocate enough space to zero-terminate
			 * the line later if necessary.
			 */
			sz = ip->ib_end - lastnl - hadnl;
			line = smalloc(sz + 1);
			memcpy(line, lastnl + hadnl, sz);
			ip->ib_cur = lastnl + hadnl;
		} else
			line = NULL;
nextbuf:
		if (ib_read(ip) == EOF) {
			if (line) {
				matchline(line, sz, sus, ip);
				free(line);
				line = NULL;
				sz = 0;
			}
			break;
		}
		ip->ib_cur--;
		if (line) {
			/*
			 * Append the partial line at the beginning of the
			 * file buffer to the line buffer.
			 */
			size_t oldsz = sz;
			if ((cp = memchr(ip->ib_cur, '\n',
					ip->ib_end - ip->ib_cur)) == NULL) {
				char	*nline;
				/*
				 * Ugh. This is really a huge line. Store the
				 * entire file buffer in the line buffer and
				 * read the next part of the file.
				 */
				sz += ip->ib_end - ip->ib_cur;
				if ((nline = realloc(line, sz + 1)) == NULL) {
					sz = oldsz;
					cp = &ip->ib_end[-1];
					oom++;
				} else {
					line = nline;
					memcpy(line + oldsz, ip->ib_cur,
						ip->ib_end - ip->ib_cur);
					goto nextbuf;
				}
			}
			if ((sz = cp - ip->ib_cur) > 0) {
				char	*nline;
				sz += oldsz;
				if ((nline = realloc(line, sz + 1)) == NULL) {
					sz = oldsz;
					oom++;
				} else {
					line = nline;
					memcpy(line + oldsz, ip->ib_cur,
							cp - ip->ib_cur);
				}
			} else
				sz = oldsz;
			if (matchline(line, sz, 1, ip))
				break;
			free(line);
			line = NULL;
			sz = 0;
			ip->ib_cur = cp + (oom == 0);
			oom = 0;
		}
	}
endgrep:
	if (!qflag && cflag) {
		if (filename && !hflag)
			printf("%s:", filename);
#ifdef	LONGLONG
		printf("%llu\n", (long long)lmatch);
#else
		printf("%lu\n", (long)lmatch);
#endif
	}
	return ip;
}

/*
 * Grep a named file.
 */
static void
fngrep(const char *fn, int level)
{
	struct iblok	*ip;
	struct stat	st;
	int	i;

	if (rflag && fn && (level ? rflag : stat)(fn, &st) == 0) {
		if (rflag != lstat) {
			for (i = 0; i < level; i++)
				if (st.st_dev == visited[i].v_dev &&
						st.st_ino == visited[i].v_ino)
					return;
			if (level >= vismax) {
				vismax += 20;
				visited = srealloc(visited, sizeof *visited *
						vismax);
			}
			visited[level].v_dev = st.st_dev;
			visited[level].v_ino = st.st_ino;
		}
	mode:	switch (st.st_mode&S_IFMT) {
#define	ignoring(t, s)	fprintf(stderr, "%s: ignoring %s %s\n", progname, t, s)
		case S_IFIFO:
			ignoring("named pipe", fn);
			return;
		case S_IFBLK:
			ignoring("block device", fn);
			return;
		case S_IFCHR:
			ignoring("block device", fn);
			return;
#ifdef	S_IFSOCK
		case S_IFSOCK:
			ignoring("socket", fn);
			return;
#endif	/* S_IFSOCK */
		case S_IFLNK:
			if (stat(fn, &st) < 0 || (st.st_mode&S_IFMT) == S_IFDIR)
				return;
			goto mode;
		default:
			break;
		case S_IFDIR: {
			char	*path;
			int	pend, psize, i;
			DIR	*df;
			struct dirent	*dp;

			if (hflag == 2)
				hflag = 0;
			if ((df = opendir(fn)) == NULL) {
				if (sflag == 0)
					fprintf(stderr, "%s: can't open "
							"directory %s\n",
							progname, fn);
				if (!qflag || status == 1)
					status = 2;
				return;
			}
			pend = strlen(fn);
			path = malloc(psize = pend + 2);
			strcpy(path, fn);
			path[pend++] = '/';
			while ((dp = readdir(df)) != NULL) {
				if (dp->d_name[0] == '.' &&
						(dp->d_name[1] == '\0' ||
					 	dp->d_name[1] == '.' &&
					 	dp->d_name[2] == '\0'))
					continue;
				i = 0;
				do {
					if (pend + i >= psize)
						path = srealloc(path,
								psize += 14);
					path[pend+i] = dp->d_name[i];
				} while (dp->d_name[i++]);
				filename = path;
				fngrep(path, level+1);
			}
			free(path);
			closedir(df);
			return;
		    }
		}
	}
	if (fn) {
		if ((ip = ib_open(fn, 0)) == NULL) {
			if (sflag == 0)
				fprintf(stderr, "%s: can't open %s\n",
						progname, fn);
			if (!qflag || status == 1)
				status = 2;
			return;
		}
	} else
		ip = ib_alloc(0, 0);
	ip = grep(ip);
	if (ip->ib_fd) {
		ib_close(ip);
		if (zflag && ip->ib_pid) {
			int	s;
			waitpid(ip->ib_pid, &s, 0);
			if (s)
				status = 2;
		}
	} else
		ib_free(ip);
}

int
main(int argc, char **argv)
{
	int i, hadpat = 0;

#ifdef	__GLIBC__
	putenv("POSIXLY_CORRECT=1");
#endif
	progname = basename(argv[0]);
	setlocale(LC_COLLATE, "");
	setlocale(LC_CTYPE, "");
	mb_cur_max = MB_CUR_MAX;
	range = gn_range;
	init();
	while ((i = getopt(argc, argv, options)) != EOF) {
		switch (i) {
		case 'E':
			Eflag |= 1;
			rc_select();
			break;
		case 'F':
			if (Eflag&2)
				Eflag = 0;
			Fflag |= 1;
			ac_select();
			break;
		case 'b':
			bflag = 1;
			break;
		case 'c':
			cflag = 1;
			break;
		case 'e':
			patstring(optarg);
			hadpat++;
			break;
		case 'f':
			fflag++;
			patfile(optarg);
			hadpat++;
			break;
		case 'H':
			Hflag = 1;
			break;
		case 'h':
			hflag = 1;
			break;
		case 'i':
		case 'y':
			iflag = 1;
			break;
		case 'l':
			lflag = 1;
			break;
		case 'n':
			nflag = 1;
			break;
		case 'q':
			qflag = 1;
			break;
		case 'r':
			rflag = stat;
			break;
		case 'R':
			rflag = lstat;
			break;
		case 's':
			sflag = 1;
			break;
		case 'v':
			vflag = 1;
			break;
		case 'w':
			wflag = 1;
			break;
		case 'x':
			xflag = 1;
			break;
		case 'z':
			zflag = 1;
			break;
		default:
			if (!(Fflag&2))
				usage();
			status = 2;
		}
	}
	if (sus) {
		if (Fflag == 2) {
			if (sflag) {
				optind = 1;
				argv[1] = "-s";
				getopt(argc, argv, "");
				usage();
			}
			if (qflag) {
				optind = 1;
				argv[1] = "-q";
				getopt(argc, argv, "");
				usage();
			}
		}
		if (Fflag && status == 2)
			usage();
		if (Eflag == 1 && Fflag == 1 || cflag + lflag + qflag > 1)
			usage();
		if (wflag && (Eflag || Fflag))
			usage();
	}
	if (cflag)
		lflag = 0;
	if (hadpat == 0) {
		if (optind >= argc)
			misop();
		patstring(argv[optind++]);
	} else if (e0 == NULL)
		patstring(NULL);
	build();
	if (optind != argc) {
		if (optind + 1 == argc)
			hflag = 2;
		do {
			if (sus && argv[optind][0] == '-' &&
					argv[optind][1] == '\0') {
				filename = NULL;
				fngrep(NULL, 0);
			} else {
				filename = argv[optind];
				fngrep(argv[optind], 0);
			}
		} while (++optind < argc);
	} else {
		if (lflag && !sus && (Eflag || Fflag))
			exit(1);
		fngrep(NULL, 0);
	}
	return status;
}
